src/index.tsx

```diff
@@ -176,14 +193,51 @@ function renderScene(
       const lineDash = context.getLineDash();
       context.setLineDash([8, 4]);
       context.strokeRect(
-        elementX1 - margin,
-        elementY1 - margin,
+        elementX1 - margin + sceneState.scrollX,
+        elementY1 - margin + sceneState.scrollY,
         elementX2 - elementX1 + margin * 2,
         elementY2 - elementY1 + margin * 2
       );
       context.setLineDash(lineDash);
     }
   });
+
+  let minX = Infinity;
+  let maxX = 0;
+  let minY = Infinity;
+  let maxY = 0;
+
+  elements.forEach(element => {
+    minX = Math.min(minX, getElementAbsoluteX1(element));
+    maxX = Math.max(maxX, getElementAbsoluteX2(element));
+    minY = Math.min(minY, getElementAbsoluteY1(element));
+    maxY = Math.max(maxY, getElementAbsoluteY2(element));
+  });
+
+  // horizontal scrollbar
+  const sceneWidth = canvasWidth + Math.abs(sceneState.scrollX);
+  const scrollBarWidth = (canvasWidth * canvasWidth) / sceneWidth;
+  const scrollBarX = sceneState.scrollX > 0 ? 0 : canvasWidth - scrollBarWidth;
+  context.fillStyle = SCROLLBAR_COLOR;
+  context.fillRect(
+    scrollBarX + SCROLLBAR_MARGIN,
+    canvasHeight - SCROLLBAR_WIDTH - SCROLLBAR_MARGIN,
+    scrollBarWidth - SCROLLBAR_MARGIN * 2,
+    SCROLLBAR_WIDTH
+  );
+
+  // vertical scrollbar
+  const sceneHeight = canvasHeight + Math.abs(sceneState.scrollY);
+  const scrollBarHeight = (canvasHeight * canvasHeight) / sceneHeight;
+  const scrollBarY =
+    sceneState.scrollY > 0 ? 0 : canvasHeight - scrollBarHeight;
+  context.fillStyle = SCROLLBAR_COLOR;
```

Comment:
can you store fillStyle and revert it after it's drawn, so it doesn't accidentally pollute the rest of the draw commands